compiler/rustc_monomorphize/src/mono_checks/abi_check.rs

```diff
@@ -88,6 +91,40 @@ fn do_check_abi<'tcx>(
     }
 }
 
+/// Warns against usage of `extern "C"` on wasm32-unknown-unknown that is affected by the
+/// ABI transition.
+fn do_check_wasm_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    is_call: bool,
+    span: impl Fn() -> Span,
+) {
+    // Only proceed for `extern "C" fn`` on wasm32-unknown-unknown (same check as what `adjust_for_foreign_abi` uses to call `compute_wasm_abi_info`).
+    if !(tcx.sess.target.arch == "wasm32"
+        && tcx.sess.target.os == "unknown"
+        && tcx.wasm_c_abi_opt() == WasmCAbi::Legacy
+        && abi.conv == Conv::C)
+    {
+        return;
+    }
+    // Warn against all types whose ABI will change. That's all arguments except for things passed as scalars.
+    // Return values are not affected by this change. (FIXME: is this true?)
+    for arg_abi in abi.args.iter() {
+        if matches!(arg_abi.layout.backend_repr, BackendRepr::Scalar(_)) {
+            continue;
+        }
```

Comment:
Structs that recursively only contain a single scalar and don't specify a larger alignment than the default are passed by-val with the new abi too.

Reply:
And 1-ZST types are ignored by both abi's too.

Reply:
ZST are generally not safe to pass to non-Rust code anyway so I won't bother with them here.

Reply:
> Structs that recursively only contain a single scalar and don't specify a larger alignment than the default are passed by-val with the new abi too.

I copied the wasm ABI adjustment logic for this.